abdiff/core/collate_ab_transforms.py

```diff
@@ -0,0 +1,220 @@
+    collated_dataset_path = str(Path(run_directory) / "collated")
+
+    transformed_written_files = write_to_dataset(
+        get_transformed_batches_iter(run_directory, transformed_files),
+        schema=TRANSFORMED_DATASET_SCHEMA,
+        base_dir=transformed_dataset_path.name,
+        partition_columns=["transformed_file_name"],
+    )
+    logger.info(
+        f"Wrote {len(transformed_written_files)} parquet file(s) to transformed dataset"
+    )
+
+    joined_written_files = write_to_dataset(
+        get_joined_batches_iter(transformed_dataset_path.name),
+        base_dir=collated_dataset_path,
+        schema=JOINED_DATASET_SCHEMA,
+    )
+    logger.info(f"Wrote {len(joined_written_files)} parquet file(s) to collated dataset")
+    return collated_dataset_path
+
+
+def yield_records(transformed_file: str | Path) -> Generator:
+    """Yields data for every TIMDEX record in a transformed file.
+
+    This function uses ijson to yield records from a JSON stream
+    (i.e., the transformed file) one at a time. A generator is returned,
+    yielding a dictionary that contains the following:
+
+    * timdex_record_id: The TIMDEX record ID.
+    * source: The shorthand name of the source as denoted in by Transmogrifier
+      (see https://github.com/MITLibraries/transmogrifier/blob/main/transmogrifier/config.py).
+    * record: The TIMDEX record serialized to a JSON string then encoded to bytes.
+    * version: The version of the transform, parsed from the absolute filepath to a
+      transformed file.
+    * transformed_file_name: The name of the transformed file, excluding file extension.
+    """
+    version, transformed_file_name = parse_parquet_details_from_transformed_file(
+        str(transformed_file)
+    )
+    with open(transformed_file) as file:
+        records = ijson.items(file, "item")
+        for record in records:
+            yield {
+                "timdex_record_id": record["timdex_record_id"],
+                "source": record["source"],
+                "record": json.dumps(record).encode(),
+                "version": version,
+                "transformed_file_name": transformed_file_name,
+            }
+
+
+def get_transformed_batches_iter(
+    run_directory: str, transformed_files: tuple[list[str], ...]
+) -> Generator:
+    """Yield pyarrow.RecordBatch objects of TIMDEX records.
+
+    This function will iterate over the A/B lists in 'transformed_files',
+    calling yield_records() to fetch dictionaries describing
+    TIMDEX records and compiling the dictionaries into a
+    pyarrow.RecordBatch. The size of a batch is set by the
+    'READ_BATCH_SIZE' global variable.
+
+    The function returns a generator, yielding batches of the dictionaries
+    from yield_records(). The returned generator can be passed to
+    abdiff.core.utils.write_to_dataset() to perform batch writes to
+    a Parquet dataset.
+    """
+    for transformed_files_list in transformed_files:
```

Comment:
Optional: `for transformed_files_list in transformed_file_lists:`, a bit of whiplash from shifting variable names in the next few lines

Reply:
I'd second this - forgot to comment on my review.  Maybe the input argument of `transformed_file_lists` could more accurately be something like `a_and_b_transformed_file_lists` (though kind of a mouthful).

Reply:
Hmm, what do ya'll think of the following?:

```python

def collate_ab_transforms(run_directory: str, ab_transformed_file_lists: tuple[list[str], ...])
   ...

def get_transformed_batches_iter(run_directory: str, ab_transformed_file_lists: tuple[list[str], ...]) -> Generator[pa.RecordBatch]:
   for transformed_files in ab_transformed_file_lists:
      for transformed_file in transformed_files:
   ...
```

**Note:** Pretty much @ghukill 's suggestion without the conjunction "and" to keep it consistent with our other function names (using "ab").

Reply:
I dig it!

Reply:
Looks good!

Reply:
Updated!